### PR TITLE
[inductor-perf] Reuse r_numel from args inputs in indexing to remove redundant kernel arguments

### PR DESCRIPTION
**Pull request by laithsakka**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #196362

While comparing static and dynamic kernels, I noticed that some indexing expressions repeatedly reconstruct a value already passed as a kernel argument.

For example, the wrapper computes and passes r0_numel:
```
r0_numel = (349 + s110) // 350

kernel.run(
    ...,
    s110,       # ks0
    r0_numel,
)
```
but the generated kernel reconstructs the same exact expression in several indices:
```
def kernel(..., ks0, r0_numel, ...):
    for r0_offset in tl.range(0, r0_numel, R0_BLOCK):
        ...
        index = r0_2 + x1 * ((349 + ks0) // 350)
```
This can instead reuse the existing argument:
```
def kernel(..., r0_numel, ...):
    for r0_offset in tl.range(0, r0_numel, R0_BLOCK):
        ...
        index = r0_2 + x1 * r0_numel
```
Replacing every occurrence produced 1–4% kernel-level wins in several benchmarks, but regressed other kernels by about 0.5%. Digging deeper, the wins consistently occurred when the replacement removed the final use of a symbolic ks* argument, allowing that argument to disappear from the kernel signature. When all original arguments remained necessary elsewhere, the rewrite was neutral or slightly slower.

This PR therefore limits the optimization to replacements that reduce the number of kernel arguments.
<details>

Reuse existing r0_numel and r1_numel kernel arguments in Triton indexing expressions when doing so removes a redundant scalar size argument from the kernel signature.

Inductor already passes reduction extents to Triton for reduction loop bounds and masks. Tensor indexing is generated separately from tensor layout and stride expressions, so the same expression can also appear as an ordinary symbolic ks* argument. This commonly happens when a reduction  is used as the stride of an outer dimension or is reconstructed from dynamic input sizes.

For example:

#### Before
```
def kernel(..., ks0, r0_numel, ...):
    for r0_offset in tl.range(0, r0_numel, R0_BLOCK):
        value = tl.load(
            in_ptr + r0_index + ks0 * xindex,
            ...
        )
```
When ks0 is exactly the same expression as r0_numel, this becomes:

#### After
```
def kernel(..., r0_numel, ...):
    for r0_offset in tl.range(0, r0_numel, R0_BLOCK):
        value = tl.load(
            in_ptr + r0_index + r0_numel * xindex,
            ...
        )
 ```
 
This removes ks0 from the kernel signature if it has no remaining uses.

### The optimization:

- Supports the ordinary r0_ and r1_ reduction families.
- Collects indexing and range-tree expressions before source emission.
- Simulates all exact reduction-extent replacements together, allowing multiple replacements to jointly eliminate an argument.
- Applies only replacements that contribute to removing a scalar kernel argument.
-0 Skips staged reductions because they can emit symbolic indexing expressions after the indexing prepass.

### Performance
The following programs and kernel classes were evaluated:

- Production model.
- Dynamic LayerNorm forward and backward standalone reproduction.
- DistilBERT dynamic training.
- Ten-model dynamic inference sweep, including DistilGPT2 and T5Small.
- Tiny and medium raw Triton reductions.
- Exact production-style ceil-div reconstruction.
- Forced cooperative reduction.
- Large 4096 x 4096 reduction.
- Three preserved LayerNorm forward/backward kernels.
- Six non-eliminating LayerNorm-backward shape configurations.
- Dynamic two-dimensional r0_/r1_ tiled reduction.
333
Representative results:

#### Production corpus:
  -   12 / 14 affected kernels improved
  - 2.69% median affected-kernel improvement
   -  1.01% combined improvement
   - 2.88% representative-kernel improvement

####  Preserved LayerNorm kernels removing ks*:
    4.73% faster
    2.33% faster

####  Raw production-style r0_numel:
    1.64-3.17% faster

####  Dynamic ND r1_numel:
    0.97% faster
    95% CI: [0.72%, 1.32%]

NOTE: why limiting only when args reduced. 
#### Forced rewrite without argument removal:
    0.47% slower
    95% CI: [0.46%, 0.48%]
    

</details>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by ezyang** [comment]
Review by **Astra agent**

**[P2] Include indirect-indexing sizes in the profitability check**

At [torch/_inductor/codegen/triton.py:3369–3372](https://github.com/pytorch/pytorch/blob/4449851577e3dd62398eceb2f87cb5f74cff1942/torch/_inductor/codegen/triton.py#L3369-L3372), `indices` and range-tree expressions omit the sizes captured by `LoopBody.bind_set_indirect_shim`. Those sizes can later require a `ks*` argument for negative-index wrapping or bounds checks.

For a reduction extent of `CeilDiv(s0, 2)`, the prepass can therefore mark `s0` eliminated even though an indirect access still needs the original `s0`. This permits rewriting without reducing the argument count—the case this PR explicitly seeks to avoid.

Please include indirect-indexing sizes in the profitability analysis and add a fused gather/reduction test that checks the final argument count.

This finding is based on static review; I could not run the tests because the review environment lacks PyTorch and CUDA. I found no correctness issue.

**Review by eellison — COMMENTED** [review]
> Preserved LayerNorm kernels removing ks*:
4.73% faster
2.33% faster

How large are these kernels ? what win does this translate to on the model level ? 

I do think passing a number of extra symbols slows down perf, especially in cat-derived kernels where there are many symbols. But this is special cased very particularly.

**Comment by laithsakka** [comment]
> > Preserved LayerNorm kernels removing ks*:
> > 4.73% faster
> > 2.33% faster
> 
> How large are these kernels ? what win does this translate to on the model level ?
> 
> I do think passing a number of extra symbols slows down perf, especially in cat-derived kernels where there are many symbols. But this is special cased very particularly.

let me update the summary with example and actually kernel code before and after so its reproducible. 
Also seems like  wins are not only from reducing args count only, but also from avoiding arithmetic compute in kernel body.

**Comment by laithsakka** [comment]
> > > Preserved LayerNorm kernels removing ks*:
> > > 4.73% faster
> > > 2.33% faster
> > 
> > 
> > How large are these kernels ? what win does this translate to on the model level ?
> > I do think passing a number of extra symbols slows down perf, especially in cat-derived kernels where there are many symbols. But this is special cased very particularly.
> 
> let me update the summary with example and actually kernel code before and after so its reproducible. Also seems like wins are not only from reducing args count only, but also from avoiding arithmetic compute in kernel body.

end to end model win is estimated by AI to be ~1-0.8%  idk if this justified tbh so i will keep in the air for now unless we found those in a really hot inference loop in the future. @eellison
